Invalid lint violation in 0.24.0 `typescript/no-unnecessary-type-assertion` rule

While most of the new violations for this rule *are* valid, in the codebase at work I discovered one case where removing the "unnecessary" assertion which tsgolint flags causes typechecking to fail. This is a bug.

The relevant code is the ramda library's `concat` call, whose overload is:

```ts
concat<T1, T2 extends T1>(list1: readonly T1[], list2: readonly T2[]): T1[];
```

Example reproduction of the issue using an equivalent method to `concat`:

```json
{
  "rules": {
    "typescript/no-unnecessary-type-assertion": "error"
  }
}
```

```json
{
  "compilerOptions": {
    "strict": true,
    "noEmit": true,
    "skipLibCheck": true
  },
  "include": [
    "repro.ts"
  ]
}
```

```ts
type Item = { id: string | number };

declare function combine<T1, T2 extends T1>(a: readonly T1[], b: readonly T2[]): T1[];

declare const items: Item[];

// The below line is a violation of `no-unnecessary-type-assertion`, but if you fix it a type error shows up.
const result = combine([{ id: 0 } as Item], items);
```

1. `tsgo` with the assertion: no errors
2. `oxlint` with the assertion: `error typescript(no-unnecessary-type-assertion): This assertion is unnecessary since the receiver accepts the original type of the expression.`
3. `tsgo` without the assertion: `error TS2345: ... Type 'string | number' is not assignable to type 'number'`

This is actually also a bug in the upstream rule, apparently: [TypeScript ESlint playground](https://typescript-eslint.io/play/#ts=6.0.3&fileType=.tsx&code=C4TwDgpgBAksEFsoF4oG8oEsAmAuKAzsAE6YB2A5lAD5RkCuCARhMVAL4DcAUN9hAGMANgENi0AGb0yA4JgD2ZKAPnNyEADwAVAIwAaKFoBMUCAA94ZbAUM6AfAAoR%2BcSOyKhIWwG0AugaYXCDcPL2M-AEp8XT8ePkFRcWVFIix4BAJ8OERY3hUyVPECeiFgFGS1MggHbwwcfAAGDigRG2yEfzTEAgjOIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

We hit this same bug independently after bumping `oxlint` `1.71 → 1.72` (which pulls in `oxlint-tsgolint` `0.23 → 0.24`) across a large TypeScript monorepo: `oxlint --fix` removed assertions that `tsgo`/`tsc` actually require, breaking typecheck in 7 packages (~230 assertions flagged in total). Adding a couple of data points that might help.

### Regression bisection → the culprit is `oxlint-tsgolint@0.24.0` (not oxlint 1.72)

Running the repro below against each pair:

| oxlint | oxlint-tsgolint | result |
| --- | --- | --- |
| 1.71.0 | 0.23.0 | ✅ no diagnostic |
| 1.72.0 | 0.23.0 | ✅ no diagnostic |
| 1.72.0 | 0.24.0 | ❌ false positive → `--fix` breaks `tsc` |

So the behavior was introduced in tsgolint `0.24.0`; oxlint 1.72 with tsgolint pinned to 0.23 is clean.

### A second minimal repro (object-literal widening via `as unknown`)

Distinct from the `ramda.concat` generic case, but the same root cause: the assertion is load-bearing because it widens the *inferred* property type that a generic factory (`T extends Loc`) then captures. Removing it is not type-preserving.

`tsconfig.json`
```json
{ "compilerOptions": { "strict": true, "noEmit": true, "skipLibCheck": true }, "include": ["src"] }
```

`.oxlintrc.json`
```json
{ "rules": { "typescript/no-unnecessary-type-assertion": "error" }, "ignorePatterns": ["node_modules/**"] }
```

`src/repro.ts`
```ts
interface Loc {
  pathname: string;
  state: unknown;
}

declare function makeMock<T extends Loc>(factory: () => T): () => T;

export const useLoc = makeMock(() => ({
  pathname: '/',
  state: undefined as unknown,
}));

export function consume(): void {
  const loc = useLoc();
  loc.state = { from: 'x' }; // requires `state` to be `unknown`
}
```

1. `tsgo` / `tsc` with the assertion → no errors.
2. `oxlint --type-aware` (1.72.0 / tsgolint 0.24.0):
   ```
   x typescript(no-unnecessary-type-assertion): This assertion is unnecessary since the receiver accepts the original type of the expression.
    10 |   state: undefined as unknown,
   ```
3. After `oxlint --fix` (removes `as unknown`) → `tsc`:
   ```
   error TS2322: Type '{ from: string; }' is not assignable to type 'undefined'.
   ```

Same thread as the `ramda.concat` case: the "receiver accepts the original type" check doesn't account for the assertion changing the *inferred* type that's captured by a generic elsewhere.

### Possibly related (same 0.24 bump)

The sibling rule `typescript/non-nullable-type-assertion-style` also produced type-incorrect autofixes for us on 0.24 — e.g. `document.querySelector('…') as HTMLElement` (whose type is `Element | null`) was rewritten to `document.querySelector('…')!`, which is `Element`, not `HTMLElement`, breaking subsequent `HTMLElement`-only usage such as `.focus()`. I haven't reduced that one to a minimal repro yet, but it appears to be the same family of regression in 0.24.

As a workaround we pinned `oxlint-tsgolint` back to `0.23.0` (keeping oxlint at 1.72).

--- a/internal/rules/no_unnecessary_type_assertion/no_unnecessary_type_assertion.go
+++ b/internal/rules/no_unnecessary_type_assertion/no_unnecessary_type_assertion.go
@@ -630,6 +630,59 @@ var NoUnnecessaryTypeAssertionRule = rule.Rule{
 			})
 		}
 
+		hasGenericInferenceParameterAtArgument := func(callOrNew *ast.Node, argIndex int, elementPath []int) bool {
+			signature := checker.Checker_getResolvedSignature(ctx.TypeChecker, callOrNew, nil, checker.CheckModeNormal)
+			if signature == nil {
+				return false
+			}
+			for signature.Target() != nil {
+				signature = signature.Target()
+			}
+			if len(signature.TypeParameters()) == 0 {
+				return false
+			}
+
+			params := checker.Signature_parameters(signature)
+			if len(params) == 0 {
+				return false
+			}
+
+			paramIndex := argIndex
+			if paramIndex >= len(params) {
+				paramIndex = len(params) - 1
+			}
+			param := params[paramIndex]
+			paramType := checker.Checker_getTypeOfSymbol(ctx.TypeChecker, param)
+			if valueDeclaration := param.ValueDeclaration; valueDeclaration != nil &&
+				valueDeclaration.Kind == ast.KindParameter &&
+				valueDeclaration.AsParameterDeclaration().DotDotDotToken != nil {
+				if typeArguments := getTypeArguments(paramType); len(typeArguments) > 0 {
+					typeArgumentIndex := 0
+					if len(typeArguments) > 1 {
+						typeArgumentIndex = min(argIndex-paramIndex, len(typeArguments)-1)
+					}
+					paramType = typeArguments[typeArgumentIndex]
+				}
+			}
+			for _, elementIndex := range elementPath {
+				var elementType *checker.Type
+				if checker.IsTupleType(paramType) {
+					typeArguments := checker.Checker_getTypeArguments(ctx.TypeChecker, paramType)
+					if len(typeArguments) > 0 {
+						elementType = typeArguments[min(elementIndex, len(typeArguments)-1)]
+					}
+				} else {
+					elementType = utils.GetNumberIndexType(ctx.TypeChecker, paramType)
+				}
+				if elementType == nil {
+					break
+				}
+				paramType = elementType
+			}
+
+			return containsTypeVariable(paramType)
+		}
+
 		genericsMismatch := func(uncast, contextual *checker.Type) bool {
 			return slices.ContainsFunc(checker.Checker_getPropertiesOfType(ctx.TypeChecker, contextual), func(prop *ast.Symbol) bool {
 				contextualSigs := checker.Checker_getSignaturesOfType(
@@ -752,6 +805,54 @@ var NoUnnecessaryTypeAssertionRule = rule.Rule{
 				ast.IsLogicalOrCoalescingAssignmentOperator(parent.AsBinaryExpression().OperatorToken.Kind)
 		}
 
+		isNestedInArrayLiteralArgumentToGenericCall := func(node *ast.Node) bool {
+			elementPath := []int{}
+			for child, current := node, node.Parent; current != nil; child, current = current, current.Parent {
+				if ast.IsFunctionExpression(current) || ast.IsArrowFunction(current) {
+					return false
+				}
+				if !ast.IsArrayLiteralExpression(current) {
+					continue
+				}
+				elementIndex := slices.IndexFunc(current.AsArrayLiteralExpression().Elements.Nodes, func(element *ast.Node) bool {
+					return element == child || ast.SkipParentheses(element) == ast.SkipParentheses(child)
+				})
+				if elementIndex == -1 {
+					continue
+				}
+				elementPath = append([]int{elementIndex}, elementPath...)
+
+				callArgument := current
+				parent := parentThroughParens(callArgument)
+				spreadOffset := 0
+				if parent != nil && ast.IsSpreadElement(parent) {
+					spreadOffset = elementIndex
+					callArgument = parent
+					parent = parentThroughParens(callArgument)
+				}
+				if parent == nil || (!ast.IsCallExpression(parent) && !ast.IsNewExpression(parent)) {
+					continue
+				}
+				if parent.TypeArguments() != nil {
+					return false
+				}
+
+				argIndex := slices.IndexFunc(parent.Arguments(), func(candidate *ast.Node) bool {
+					return candidate == callArgument || ast.SkipParentheses(candidate) == callArgument
+				})
+				if argIndex == -1 {
+					continue
+				}
+
+				parameterElementPath := elementPath
+				if ast.IsSpreadElement(callArgument) {
+					parameterElementPath = parameterElementPath[1:]
+				}
+				return hasGenericInferenceParameterAtArgument(parent, argIndex+spreadOffset, parameterElementPath)
+			}
+			return false
+		}
+
 		isInGenericContext := func(node *ast.Node) bool {
 			seenFunction := false
 			for current := node.Parent; current != nil; current = current.Parent {
@@ -803,6 +904,7 @@ var NoUnnecessaryTypeAssertionRule = rule.Rule{
 
 			if skipParentTypeForContextualAny(node) ||
 				ast.IsArrayLiteralExpression(node.Expression()) ||
+				isNestedInArrayLiteralArgumentToGenericCall(node) ||
 				isInDestructuringDeclaration(node) ||
 				isPropertyInProblematicContext(node) ||
 				isAssignmentInNonStatementContext(node) ||

--- a/internal/rules/no_unnecessary_type_assertion/no_unnecessary_type_assertion_test.go
+++ b/internal/rules/no_unnecessary_type_assertion/no_unnecessary_type_assertion_test.go
@@ -1113,6 +1113,74 @@ const contextualValue: unknown = callable as unknown;
 declare function consume(value: unknown): void;
 consume(callable as unknown);
     `},
+		{
+			// https://github.com/oxc-project/tsgolint/issues/1044
+			Code: `
+type Item = { id: string | number };
+
+declare function combine<T1, T2 extends T1>(a: readonly T1[], b: readonly T2[]): T1[];
+
+declare const items: Item[];
+
+const result = combine([{ id: 0 } as Item], items);
+    `,
+		},
+		{
+			Code: `
+type Item = { id: string | number };
+
+interface Combiner {
+  new <T1, T2 extends T1>(a: readonly T1[], b: readonly T2[]): unknown;
+}
+
+declare const Combiner: Combiner;
+declare const items: Item[];
+
+const result = new Combiner([{ id: 0 } as Item], items);
+    `,
+		},
+		{
+			Code: `
+type Item = { id: string | number };
+
+declare function combine<T1, T2 extends T1>(
+  ...args: [label: string, a: readonly T1[], b: readonly T2[]]
+): T1[];
+
+declare const items: Item[];
+
+const result = combine('items', [{ id: 0 } as Item], items);
+    `,
+		},
+		{
+			Code: `
+type Item = { id: string | number };
+
+declare function combine<T>(
+  enabled: boolean,
+  ...args: [a: readonly T[], b: readonly Item[]]
+): T[];
+
+declare const items: Item[];
+
+const result = combine(true, [{ id: 0 } as Item], items);
+result[0].id = 'x';
+    `,
+		},
+		{
+			Code: `
+type Item = { id: string | number };
+
+declare function combine<T1, T2 extends T1>(
+  ...args: [a: readonly T1[], b: readonly T2[]]
+): T1[];
+
+declare const items: Item[];
+
+const result = combine(...[[{ id: 0 } as Item], items]);
+result[0].id = 'x';
+    `,
+		},
 	}, []rule_tester.InvalidTestCase{
 		{
 			Code:   "const foo = <3>3;",
@@ -3265,5 +3333,50 @@ const value: unknown = 42;`},
 				{MessageId: "contextuallyUnnecessary"},
 			},
 		},
-	})
+		{
+			Code:   "declare function f<T>(xs: string[], tag: T): void;\nf(['x' as string], 1);",
+			Output: []string{"declare function f<T>(xs: string[], tag: T): void;\nf(['x'], 1);"},
+			Errors: []rule_tester.InvalidTestCaseError{
+				{
+					MessageId: "contextuallyUnnecessary",
+				},
+			},
+		},
+		{
+			Code:   "type StringsWithMeta<T> = string[] & { meta?: T };\ndeclare function f<T>(xs: StringsWithMeta<T>, tag: T): void;\nf(['x' as string], 1);",
+			Output: []string{"type StringsWithMeta<T> = string[] & { meta?: T };\ndeclare function f<T>(xs: StringsWithMeta<T>, tag: T): void;\nf(['x'], 1);"},
+			Errors: []rule_tester.InvalidTestCaseError{
+				{
+					MessageId: "contextuallyUnnecessary",
+				},
+			},
+		},
+		{
+			Code: `
+declare function f(xs: string[]): void;
+declare function f<T>(value: T): void;
+f(['x' as string]);`,
+			Output: []string{`
+declare function f(xs: string[]): void;
+declare function f<T>(value: T): void;
+f(['x']);`},
+			Errors: []rule_tester.InvalidTestCaseError{
+				{
+					MessageId: "contextuallyUnnecessary",
+				},
+			},
+		},
+		{
+			Code: `
+declare function f<T>(pair: readonly [T, string], tag: T): void;
+f([1, 'x' as string], 1);`,
+			Output: []string{`
+declare function f<T>(pair: readonly [T, string], tag: T): void;
+f([1, 'x'], 1);`},
+			Errors: []rule_tester.InvalidTestCaseError{
+				{
+					MessageId: "contextuallyUnnecessary",
+				},
+			},
+		}})
 }

